tools/train.py

```diff
@@ -62,6 +68,81 @@ def parse_args():
     return args
 
 
+def determine_max_batch_size(cfg, distributed):
+    def get_fake_input(cfg, orig_img_shape=(128, 128, 3), device='cuda'):
+        test_pipeline = [LoadImage()] + cfg.data.test.pipeline[1:]
+        test_pipeline = Compose(test_pipeline)
+        data = dict(img=np.zeros(orig_img_shape, dtype=np.uint8))
+        data = test_pipeline(data)
+        data = scatter(collate([data], samples_per_gpu=1), [device])[0]
+        return data
+
+    model = build_detector(
+        cfg.model, train_cfg=cfg.train_cfg, test_cfg=cfg.test_cfg).cuda()
+
+    if 'pipeline' in cfg.data.train:
+        img_shape = [t for t in cfg.data.train.pipeline if t['type'] == 'Resize'][0]['img_scale']
+    else:
+        img_shape = [t for t in cfg.data.train.dataset.pipeline if t['type'] == 'Resize'][0][
+            'img_scale']
+
+    channels = 3
+
+    fake_input = get_fake_input(cfg, orig_img_shape=list(img_shape) + [channels])
+    img_shape = fake_input['img_metas'][0][0]['pad_shape']
+
+    width, height = img_shape[0], img_shape[1]
+
+    min_bs = 2
+    max_bs = 512
+    step = 2
+
+    batch_size = min_bs
+    for bs in range(min_bs, max_bs, step):
+        try:
+            gt_boxes = [torch.tensor([[0., 0., width, height]]).cuda() for _ in range(bs)]
+            gt_labels = [torch.tensor([0], dtype=torch.long).cuda() for _ in range(bs)]
+            img_metas = [fake_input['img_metas'][0][0] for _ in range(bs)]
+
+            gt_masks = None
+
+            if isinstance(model, TwoStageDetector):
+                if model.roi_head.with_mask:
+                    rles = maskUtils.frPyObjects(
+                        [[0.0, 0.0, width, 0.0, width, height, 0.0, height]], height, width)
+                    rle = maskUtils.merge(rles)
+                    mask = maskUtils.decode(rle)
+                    gt_masks = [BitmapMasks([mask], height, width) for _ in range(bs)]
+
+            if gt_masks is None:
+                model(torch.rand(bs, channels, height, width).cuda(), img_metas=img_metas,
+                      gt_bboxes=gt_boxes, gt_labels=gt_labels)
+            else:
+                model(torch.rand(bs, channels, height, width).cuda(), img_metas=img_metas,
+                      gt_bboxes=gt_boxes, gt_labels=gt_labels, gt_masks=gt_masks)
+
+            batch_size = bs
```

Comment:
never tried, ok, will check

Reply:
What do you mean by that? If you use non-distributed training you can use one gpu only, is not it?

Reply:
Nope. Distributed training allows you to run training on multiple machines, when non-distributed one restricts you to use only one. But in both cases you may have multiple GPUs on every machine.

Reply:
ok, in that case I did not test it in distributed mode. In case of non-distributed training with multiple gpus it works well. Batch size is number of samples per gpu.

Reply:
Have no idea how you even managed to run this in a non-distributed setup. According to [this](https://github.com/open-mmlab/mmdetection/issues/3116) non-distributed training support unfortunately has been dropped.